app/models/miq_filter.rb

```diff
@@ -43,28 +43,43 @@ def self.belongsto2object_list(tag)
     # /belongsto/ExtManagementSystem|<name>/EmsCluster|<name>/EmsFolder|<name>
     raise _("invalid tag: %{tag}") % {:tag => tag} unless tag.starts_with?("/belongsto/ExtManagementSystem")
     parts = tag.split("/")[2..-1]
+    depth = parts.size - 1 # ancestry uses 0 based depth
 
-    # root object
-    klass, name = parts.shift.split("|")
-    obj = find_object_by_name(klass, name)
+    # Get the root EMS object
+    # TODO: For tree queries deeper than 1, we don't actually need the ems object,
+    #       so find a way to just get the id
+    ems_class, ems_name = parts.first.split("|", 2)
+    ems = find_object_by_name(ems_class, ems_name)
 
-    if obj.nil?
-      _log.warn("lookup for klass=#{klass.to_s.inspect} with name=#{name.inspect} failed in tag=#{tag.inspect}")
+    if ems.nil?
+      _log.warn("lookup for klass=#{ems_class.to_s.inspect} with name=#{ems_name.inspect} failed in tag=#{tag.inspect}")
       return []
     end
 
-    # traverse the tree
-    parts.each_with_object([obj]) do |p, result|
-      tag_part_klass, name = p.split("|")
-      tag_part_klass = tag_part_klass.constantize
+    return [ems] if depth == 0
 
-      obj = obj.with_relationship_type('ems_metadata') do
-        obj.children.grep(tag_part_klass).detect { |c| c.name == name }
-      end
+    # Get the leaf node object for this EMS
+    leaf_class, leaf_name = parts.last.split("|", 2)
+    leaves = leaf_class.constantize
+      .includes(:all_relationships)
+      .where(:name => leaf_name, :ems_id => ems.id)
```

Comment:
probably as a separate call, is it possible to bring back the relationship records for all of these.

that will keep the query count relatively constant and will get us the depth, string path, and result_parts

we do need to be aware that some of these leaf nodes may have multiple paths. not sure if we deal 100% in paths if this would alleviate our issues.

Reply:
Isn't that what `.includes(:all_relationships)` does?

Reply:
> we do need to be aware that some of these leaf nodes may have multiple paths. not sure if we deal 100% in paths if this would alleviate our issues.

that's an interesting point - I don't think VMs can be belongsto records, and I'm pretty sure VMs are the only things that can have multiple paths.

Reply:
can we put in a `where(:relationship_type =>)` to limit the relationships?

Reply:
yes, great point - let me add that in.

Reply:
Updated.